src/types/transactions.rs

```diff
@@ -195,14 +204,77 @@ pub struct BakerKeysPayload {
     pub proof_aggregation:      aggregate_sig::Proof<AggregateSigPairing>,
 }
 
+impl<T> BakerKeysPayload<T> {
+    /// Construct a BakerKeysPayload taking a prefix for the challenge.
+    fn new<R: CryptoRng + RngCore>(
+        csprng: &mut R,
+        baker_keys: &BakerKeyPairs,
+        sender: AccountAddress,
+        challenge_prefix: &[u8],
+    ) -> Self {
+        let mut challenge = challenge_prefix.to_vec();
+
+        sender.serial(&mut challenge);
+        baker_keys.election_verify.serial(&mut challenge);
+        baker_keys.signature_verify.serial(&mut challenge);
+        baker_keys.aggregation_verify.serial(&mut challenge);
+
+        let proof_election = eddsa_ed25519::prove_dlog_ed25519(
+            &mut RandomOracle::domain(&challenge),
+            &baker_keys.election_verify.verify_key,
+            &baker_keys.election_sign.sign_key,
+        );
+        let proof_sig = eddsa_ed25519::prove_dlog_ed25519(
+            &mut RandomOracle::domain(&challenge),
+            &baker_keys.signature_verify.verify_key,
+            &baker_keys.signature_sign.sign_key,
+        );
+        let proof_aggregation = baker_keys
+            .aggregation_sign
+            .prove(csprng, &mut RandomOracle::domain(&challenge));
+
+        BakerKeysPayload {
+            phantom: PhantomData::default(),
+            election_verify_key: baker_keys.election_verify.clone(),
+            signature_verify_key: baker_keys.signature_verify.clone(),
+            aggregation_verify_key: baker_keys.aggregation_verify.clone(),
+            proof_sig,
+            proof_election,
+            proof_aggregation,
+        }
+    }
+}
+
+impl BakerKeysPayload<AddBakerKeysMarker> {
+    /// Construct a BakerKeysPayload with proofs for adding a baker.
+    pub fn new_add_payload<T: CryptoRng + RngCore>(
+        csprng: &mut T,
```

Comment:
This argument should always be last, to match what is there for encrypted transfers.

There is still some inconsistency left with this, but we should strive to eliminate it.